src/persist-client/src/internal/encoding.rs

```diff
@@ -189,19 +189,26 @@ pub(crate) fn parse_id(id_prefix: char, id_type: &str, encoded: &str) -> Result<
 // decode time, we're able to compare the current version against any we receive
 // and assert as necessary.
 //
-// Initially we reject any version from the future (no forward compatibility,
-// most conservative but easiest to reason about) but allow any from the past
-// (permanent backward compatibility). If/when we support deploy rollbacks and
-// rolling upgrades, we can adjust this assert as necessary to reflect the
-// policy (e.g. by adding some window of X allowed versions of forward
-// compatibility, computed by comparing semvers).
+// Initially we allow any from the past (permanent backward compatibility) and
+// one minor version into the future (forward compatibility). This allows us to
+// run two versions concurrently for rolling upgrades. We'll have to revisit
+// this logic if/when we start using major versions other than 0.
 //
 // We could do the same for blob data, but it shouldn't be necessary. Any blob
 // data we read is going to be because we fetched it using a pointer stored in
 // some persist state. If we can handle the state, we can handle the blobs it
 // references, too.
 pub(crate) fn check_data_version(code_version: &Version, data_version: &Version) {
```

Comment:
This method was also recently reused for checking the versions of batches about to be CaA'd. AFAIK allowing forward compatibility on those isn't necessary for unblocking the persist catalog impl, so we could fork this method and have different logic.

I didn't do that here because the surface area is small compared to all of persist state, and with persist-txn feels like 0dt is probably going to need it anyway? Even if not, probably surprising for one to be forward compatible and one not.

All that said, I'm definitely amenable to an argument in favor of keeping the minimum possible forward compatible surface area if anyone has that opinion here.